### PR TITLE
add overriding condition to BackendTrafficPolicy and SecurityPolicy

### PR DESCRIPTION
**Pull request by zhaohuabing**
fix https://github.com/envoyproxy/gateway/issues/2055

### CODE REVIEW

File: internal/gatewayapi/backendtrafficpolicy.go [hunk 1]
```diff
@@ -87,6 +92,32 @@ func (t *Translator) ProcessBackendTrafficPolicies(backendTrafficPolicies []*egv
 				continue
 			}
 
+			// Populate the gatewayRouteMap that will be used to check policy overrides
+			var parents []gwapiv1.ParentReference
+			switch r := route.(type) {
```
Comment (arkodg) [inline internal/gatewayapi/backendtrafficpolicy.go:97]: can use https://github.com/envoyproxy/gateway/blob/72fadb7c1dbe7d660c2900527c34f53cfcfce7ac/internal/gatewayapi/contexts.go#L216 here

Comment (arkodg) [reply internal/gatewayapi/backendtrafficpolicy.go:97]: or even better can use https://github.com/envoyproxy/gateway/blob/72fadb7c1dbe7d660c2900527c34f53cfcfce7ac/internal/gatewayapi/contexts.go#L242 directly

File: internal/gatewayapi/backendtrafficpolicy.go [hunk 2]
```diff
@@ -108,6 +139,23 @@ func (t *Translator) ProcessBackendTrafficPolicies(backendTrafficPolicies []*egv
 
 			t.translateBackendTrafficPolicyForGateway(policy, gateway, xdsIR)
 
+			// Check if this policy is overridden by other policies
```
Comment (arkodg) [inline internal/gatewayapi/backendtrafficpolicy.go:142]: shouldnt this be executed before `translateBackendTrafficPolicyForGateway`

**Comment by arkodg** [comment]
thanks for picking this one up @zhaohuabing 
* can the PR also apply the same logic for SecurityPolicy ?
* as part of this PR can we also fix the logic here 
https://github.com/envoyproxy/gateway/blob/72fadb7c1dbe7d660c2900527c34f53cfcfce7ac/internal/gatewayapi/backendtrafficpolicy.go#L364
as pointed out by you earlier, its adoping `Merge` semantics, when this should be a `Replace` i.e. if that route has an existing policy attached (maybe computed using the `routeMap`, we should skip applying any setting, rather than checking which is `nil`)

**Comment by zhaohuabing** [comment]
> can the PR also apply the same logic for SecurityPolicy ?

Yes, I originally planned to apply this to `SecurityPolicy` when this one is approved.  But since you already mentioned this, I'm going to do it in this PR.

> as pointed out by you earlier, its adoping Merge semantics, when this should be a Replace i.e. if that route has an existing policy attached (maybe computed using the routeMap, we should skip applying any setting, rather than checking which is nil)

I thought it was intentional. For example, a default global config can be defined at the Gateway level, and app developers can fine-tune the configurations at the route level. Having a global config on the top, the app developers don't have to redefine every setting at the roue level, but only these they really want to modify.

**Comment by arkodg** [comment]
> I thought it was intentional. For example, a default global config can be defined at the Gateway level, and app developers can fine-tune the configurations at the route level.

The benefit of having a global config on the top is that the app developers don't have to redefine every setting at the roue level, but only these they really want to modify. The downside is that it's difficult to figure out wich setting comes from which policy.

@zhaohuabing `Replace` by the most specific resource is the default https://gateway.envoyproxy.io/v0.6.0/design/security-policy/#design-decisions , and support for `Merge` will hopefully be tackled in the near future with https://github.com/envoyproxy/gateway/issues/1934

**Comment by zhaohuabing** [comment]
> @zhaohuabing Replace by the most specific resource is the default https://gateway.envoyproxy.io/v0.6.0/design/security-policy/#design-decisions , and support for Merge will hopefully be tackled in the near future with https://github.com/envoyproxy/gateway/issues/1934

@arkodg Got it! Do you mind if I fix this in a separate PR?

**Comment by arkodg** [comment]
> > @zhaohuabing Replace by the most specific resource is the default https://gateway.envoyproxy.io/v0.6.0/design/security-policy/#design-decisions , and support for Merge will hopefully be tackled in the near future with #1934
> 
> @arkodg Got it! Do you mind if I fix this in a separate PR?

Sure sgtm !

**Review by arkodg — APPROVED** [review]
LGTM thanks !

**Pull request merged**